src/DataLib/fendl32B_retrofit/tendl_preprocessing.py

```diff
@@ -0,0 +1,319 @@
+    return temp_file
+
+def download_tendl(element, A, filetype, save_path = None):
+    """
+    Download ENDF/PENDF files from the TENDL 2017 database for specific isotopes.
+
+    Arguments:
+        element (str): Chemical symbol for element of interest (i.e. Ti).
+        A (str or int): Mass number for selected isotope (i.e. 48).
+            If the target is an isomer, "m" after the mass number (i.e. 48m),
+            so A must be input as a string.
+        filetype (str): Either "ENDF" or "PENDF" filetypes to be downloaded
+            (case insensitive).
+        save_path (str, optional): User-defined file path for the downloaded file.
+            Defaults to None and will be otherwise defined internally.
+    
+    Returns:
+        save_path (str): File path for the file downloaded file.
+    """
+
+    # Ensure that A is properly formatted
+    A = A.zfill(4) if 'm' in A else str(A).zfill(3)
+
+    # Create a dictionary to generalize formatting for both ENDF and PENDF files
+    file_handling = {'endf' : {'ext': 'tendl', 'tape_num': 20},
+                     'pendf' : {'ext': 'pendf', 'tape_num': 21}}
+    
+    # Construct the filetype and isotope specific URL
+    isotope_component = f'{element}/{element}{A}/lib/endf/n-{element}{A}.'
+    ext = file_handling[filetype.lower()]['ext']
+    download_url = TENDL_GEN_URL + isotope_component + ext
+    logger.info(f'{filetype.upper()} URL: {download_url}')
+
+    # Define a save path for the file if there is not one already specified
+    if save_path is None:
+        save_path = f'tape{file_handling[filetype.lower()]["tape_num"]}'
+
+    # Conditionally downloa
+    temp_file = urllib_download(download_url, filetype)
+
+    with open(save_path, 'w') as f:
+        f.write(temp_file)
+
+    return save_path
+
+def extract_gendf_pkza(gendf_path, M=None):
+    """
+    Read in and parse the contents of a GENDF file to construct the parent KZA.
+        KZA values are defined as ZZAAAM, where ZZ is the isotope's atomic number,
+        AAA is the mass number, and M is the isomeric state (0 if non-isomeric).
+    
+    Arguments:
+        gendf_path (str): File path to the GENDF file being analyzed.
+        M (str, optional): Identifier of isomer, signified by the letter "M" at the end
+            of the mass number string.
+            Defaults to None and will be otherwise defined internally.
+    
+    Returns:
+        pKZA (str): Parent KZA identifier.
+    """
+
+    with open(gendf_path, 'r') as f:
+        first_line = f.readline()
+    logger.info(f"First line of GENDF file: {first_line}")
+    Z, element, A = first_line.split('-')[:3]
+    A = A.split(' ')[0]
+    if 'm' in A:
+        m_index = A.find('m')
+        A = A[:m_index]
+    M = str(str(M).count('M')) or '0'
+    pKZA = Z.zfill(2) + A.zfill(3) + M
+    return pKZA
+
+@contextlib.contextmanager
+def redirect_ENDFtk_output():
```

Comment:
See how things work in avoiding using this function and letting the ENDFtk messages come out organically.